diag_manager/fms_diag_object.F90

```diff
@@ -0,0 +1,368 @@
+module fms_diag_object_mod
+!> \author Tom Robinson
+!> \email thomas.robinson@noaa.gov
+!! \brief Contains routines for the diag_objects
+!!
+!! \description The diag_manager passes an object back and forth between the diag routines and the users.
+!! The procedures of this object and the types are all in this module.  The fms_dag_object is a type 
+!! that contains all of the information of the variable.  It is extended by a type that holds the
+!! appropriate buffer for the data for manipulation.
+use diag_data_mod,  only: diag_null
+use diag_data_mod,  only: r8, r4, i8, i4, string, null_type_int
+use diag_data_mod,  only: diag_null, diag_not_found, diag_not_registered, diag_registered_id
+use diag_data_mod,  only: diag_fields_type, diag_files_type
+use fms_diag_yaml_mod, only: is_field_type_null
+use fms_diag_yaml_mod, only: diag_yaml 
+use diag_axis_mod,  only: diag_axis_type
+use mpp_mod, only: fatal, note, warning, mpp_error
+!use diag_util_mod,  only: int_to_cs, logical_to_cs
+!USE diag_data_mod, ONLY: fileobjU, fileobj, fnum_for_domain, fileobjND
+
+use fms2_io_mod
+use iso_c_binding
+
+implicit none
+
+!> \brief Object that holds all variable information
+type fms_diag_object
+     type (diag_fields_type)                           :: diag_field         !< info from diag_table
+     type (diag_files_type),allocatable, dimension(:)  :: diag_file          !< info from diag_table
+     integer, allocatable, private                    :: diag_id           !< unique id for varable
+     class(FmsNetcdfFile_t), dimension (:), pointer   :: fileob => NULL()  !< A pointer to all of the 
+                                                                           !! file objects for this variable
+     character(len=:), allocatable, dimension(:)      :: metadata          !< metedata for the variable
+     logical, private                                 :: static         !< true is this is a static var
+     logical, allocatable, private                    :: registered     !< true when registered
+     integer, allocatable, dimension(:), private      :: frequency         !< specifies the frequency
+
+     integer,          allocatable, private           :: vartype           !< the type of varaible
+     character(len=:), allocatable, private           :: varname           !< the name of the variable     
+     character(len=:), allocatable, private           :: longname          !< longname of the variable     
+     character(len=:), allocatable, private           :: units             !< the units
+     character(len=:), allocatable, private           :: modname           !< the module
+     integer, private                                 :: missing_value     !< The missing fill value
```

Comment:
Is there going to be a fms_diag_object for every variable? or how will this work?

Reply:
Yes, there will be an object for every variable.  The object will be created when registered.  The data buffer will be created on the first send data call.  A linked list or something similar will manage their locations instead of a giant array of objects.